Added validate_tcp_client_information server setting

Demilivor opened the pull request:
Client info validation was added by #53673. This PR makes client info validation optional.

`validate_tcp_client_information` determines whether validation of client information enabled when query packet is received from a client using a TCP connection.

If `true`, an exception will be thrown on invalid client information from the TCP client. clickhouse-clients using version 23.7 and older are not compatible with the newer 23.8+ server version.

If `false`, the data will not be validated. The server will work with clients of all versions.

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added server setting validate_tcp_client_information determines whether validation of client information enabled when query packet is received.

### Documentation entry for user-facing changes
- [x] Documentation is written (mandatory for new features)

Demilivor committed:
Added validate_tcp_client_information setting

Demilivor commented on src/Core/ServerSettings.h:
```diff
@@ -96,7 +96,8 @@ namespace DB
     M(UInt64, total_memory_profiler_sample_min_allocation_size, 0, "Collect random allocations of size greater or equal than specified value with probability equal to `total_memory_profiler_sample_probability`. 0 means disabled. You may want to set 'max_untracked_memory' to 0 to make this threshold to work as expected.", 0) \
-    M(UInt64, total_memory_profiler_sample_max_allocation_size, 0, "Collect random allocations of size less or equal than specified value with probability equal to `total_memory_profiler_sample_probability`. 0 means disabled. You may want to set 'max_untracked_memory' to 0 to make this threshold to work as expected.", 0)
+    M(UInt64, total_memory_profiler_sample_max_allocation_size, 0, "Collect random allocations of size less or equal than specified value with probability equal to `total_memory_profiler_sample_probability`. 0 means disabled. You may want to set 'max_untracked_memory' to 0 to make this threshold to work as expected.", 0) \
+    M(Bool, validate_tcp_client_information, true, "Validate client_information in the query packet over the native TCP protocol. ClickHouse-client's with versions 23.7 can't send queries to newer servers with `validate_tcp_client_information` enabled.", 0)
```
NOTE:  
This option is enabled by default. I'm not absolutely sure that it is better than the disabled option. But I would prefer maximum security enabled by default.

The upgrade process should include simultaneous upgrades of client and server, like this:

```bash
$ sudo apt-get update
$ sudo apt-get install clickhouse-client clickhouse-server
$ sudo service clickhouse-server restart
```

On all ClickHouse hosts.

--- --- --- 
with disabled  validate_tcp_client_information by default option we also need to make tests  `02865_tcp_proxy_query_packet_validation` pass. 
To do that:
Create an XML tests/config/config.d/validate_tcp_client_information.xml file:
```xml
<clickhouse>
    <validate_tcp_client_information>true</validate_tcp_client_information>
</clickhouse>
```

And include this file in tests/config/install.sh file:
```bash
ln -sf $SRC_PATH/config.d/validate_tcp_client_information.xml $DEST_SERVER_PATH/config.d/
```

Demilivor replied:
Update:
It was decided to disable client information validation by default because there are many libraries that already use the native protocol: https://clickhouse.com/docs/en/interfaces/third-party/client-libraries

tavplubix replied:
I think we will enable it by default in the next LTS (24.3)

tavplubix replied:
Yes, it is false for now

Demilivor commented:
@tavplubix Please be the reviewer of this PR and add the label `Can Be Tested`.

Demilivor committed:
disabled validate_tcp_client_information

robot-clickhouse-ci-1 commented:
<!-- automatic status comment for PR #53907 from arenadata/ClickHouse:ADQM-1126 -->
This is an automated comment for commit de1ed978764558b0d3ee91424fc16b2537230277 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/53907/de1ed978764558b0d3ee91424fc16b2537230277/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

tavplubix commented on docs/en/operations/settings/settings.md:
```diff
@@ -4643,3 +4643,19 @@ SELECT toFloat64('1.7091'), toFloat64('1.5008753E7') SETTINGS precise_float_pars
+
+Determines whether validation of client information enabled when query packet is received from a client using a TCP connection.
+
+If `true`, an exception will be thrown on invalid client information from the TCP client. clickhouse-clients using version 23.7 and older are not compatible with the newer 23.8+ server version.
```
Let's make them compatible. Currently, you can use `clickhouse-client` version 18.1 to connect to `clickhouse-server` version 23.8 and vice versa. Let's keep it this way. We can do it by allowing both `ClickHouse` and `ClickHouse client` and converting one to another for logging purposes (as discussed in https://github.com/ClickHouse/ClickHouse/pull/53673)

Demilivor replied:
done

tavplubix reviewed (approved):
Everything else LGTM

Demilivor committed:
backward compatibility changes + fixes

Demilivor commented:
PR Update:
1) Even with `validate_tcp_client_information` enabled, old clickhouse clients (sending "ClickHouse" or "ClickHouse " as client_name) should be able to work with the new ClickHouse server.
2) Fixed the previously non-running integration test ` test_composable_protocols`.
3) Added new test cases for backward compatibility check.

Demilivor committed:
minor corrections

Demilivor committed:
minor improvement

Felixoid commented:
> sending "ClickHouse" or "ClickHouse " as client_name

AFAIS, there are no tests for such behavior. Only new `21ClickHouse client`

Besides that, the test does not [run](https://play.clickhouse.com/play?user=play#U0VMRUNUICogRlJPTSBjaGVja3MgCldIRVJFIDEKICAgIEFORCBwdWxsX3JlcXVlc3RfbnVtYmVyID0gNTM5MDcKICAgIEFORCB0ZXN0X25hbWUgSU4gKCd0ZXN0X2NvbXBvc2FibGVfcHJvdG9jb2xzL3Rlc3QucHk6OnRlc3RfY29ubmVjdGlvbnMnLCAndGVzdF9jb25jdXJyZW50X2JhY2t1cHNfczMvdGVzdC5weTo6dGVzdF9jb25jdXJyZW50X2JhY2t1cHMnKQpPUkRFUiBCWSBjaGVja19zdGFydF90aW1l) still. And to run it, we have chicken-egg issue with #54043 

I've disabled both tests in the PR to have it merged into the master finally. So it will be in the same state as it's now. And it needs to be enabled after it's fixed.

Felixoid commented:
> I added 2 more test cases

To tests/queries/0_stateless/01601_proxy_protocol.sh?

My point, it is only `21ClickHouse client` in tests/integration/test_composable_protocols/test.py now

Demilivor commented:
> To tests/queries/0_stateless/01601_proxy_protocol.sh?

Yes.

> My point, it is only 21ClickHouse client in tests/integration/test_composable_protocols/test.py now

I made only 1 change in the test:
Now it sends "ClickHouse Client" instead of "ClickHouse " in the query packet in the `client_name` field.

> Nevertheless, none of my points are blocking the merge, except it should be tested locally for correct work because now the integration test is not running in the CI

I checked every changed test, including `test_composable_protocols` on my local machine, all tests are pass.

tavplubix commented:
Stateless tests (release, wide parts enabled) - https://github.com/ClickHouse/ClickHouse/pull/54063

alexey-milovidov commented:
TLDR, we've added a setting that no one will ever enable. Let's delete the whole feature altogether.

tavplubix commented:
> @tavplubix TLDR, we've added a setting that no one will ever enable

@alexey-milovidov, please read the whole discussion. We will enable it in the next LTS

> Let's delete the whole feature altogether.

Well, then we will have to remove `session_log` as well

alexey-milovidov commented:
Ok, never mind.
